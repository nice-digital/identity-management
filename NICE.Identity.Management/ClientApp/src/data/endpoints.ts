const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
