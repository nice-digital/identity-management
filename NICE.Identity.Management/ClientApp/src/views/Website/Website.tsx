import React, { Component } from "react";

import { RouteComponentProps, Link } from "react-router-dom";

import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { PageHeader } from "@nice-digital/nds-page-header";
import { queryStringToObject, appendQueryParameter, removeQueryParameter, stripMultipleQueries } from "../../utils/querystring";
import { Table } from "@nice-digital/nds-table";
import { Grid, GridItem } from "@nice-digital/nds-grid";

import { fetchData } from "../../helpers/fetchData";
import { isDataError } from "../../helpers/isDataError";
import { RoleType, UserAndRolesType, WebsiteUsersAndRolesType } from "../../models/types";
import { Endpoints } from "../../data/endpoints";
import { FilterBox } from "../../components/FilterBox/FilterBox";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { Pagination } from "../../components/Pagination/Pagination";

import styles from "./Website.module.scss";

type TParams = { id: string };

type WebsiteProps = Record<string, unknown> & RouteComponentProps<TParams>;

type WebsiteState = {
	path: string;
	websiteUsersAndRoles: WebsiteUsersAndRolesType;
	usersAndRoles: Array<UserAndRolesType>;
	allRoles: Array<RoleType>;
	error?: Error;
	redirect: boolean;
	isLoading: boolean;
	pageNumber: number;
	itemsPerPage: number | string;
	rolesForFilter: Array<string>;
	roleFiltersChecked: Array<string>;
};

export class Website extends Component<WebsiteProps, WebsiteState> {
    constructor(props: WebsiteProps) {
		super(props);

		const querystring = this.props.location.search;

		const querystringObject = queryStringToObject(querystring);

		const pageNumber = Number(
			querystringObject.page
				? Array.isArray(querystringObject.page)
					? querystringObject.page[0]
					: querystringObject.page
				: 1,
		);

		let itemsPerPage = querystringObject.amount
			? Array.isArray(querystringObject.amount)
				? querystringObject.amount[0]
				: querystringObject.amount
			: 25;

		itemsPerPage = Number(itemsPerPage) ? Number(itemsPerPage) : itemsPerPage;

		this.state = {
			path: "",
			websiteUsersAndRoles: {} as WebsiteUsersAndRolesType,
			usersAndRoles: [],
			allRoles: [],
			redirect: false,
			isLoading: true,
			pageNumber: pageNumber,
			itemsPerPage: itemsPerPage,
			rolesForFilter: [],
			roleFiltersChecked: []
		};
	}

    async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const websiteUsersAndRoles = await fetchData(Endpoints.usersAndRolesByWebsite(this.props.match.params.id));

		if (isDataError(websiteUsersAndRoles)) {
			this.setState({ error: websiteUsersAndRoles });
		} else {
			const allRoles: Array<string> = []
			websiteUsersAndRoles.allRoles
				.slice(0)
				.map((role: RoleType) => {
					allRoles.push(role.name)
				});	
				
			
			const rolesForFilter = allRoles.reduce(function (
				accumulatedRoles: string[],
				currentRoles: string,
			) {
				if (accumulatedRoles.indexOf(currentRoles) === -1) {
					accumulatedRoles.push(currentRoles);
				}
				return accumulatedRoles;
			},
			[]);

			rolesForFilter.sort();
			this.setState({ rolesForFilter });

			document.title = `NICE Accounts - ${websiteUsersAndRoles.website.host}`;
		}

		this.setState({ websiteUsersAndRoles, usersAndRoles: websiteUsersAndRoles.usersAndRoles, isLoading: false });
	}

	getListOfUsersByFilteredRole = (
		roleFiltersChecked: Array<string>,
		usersAndRoles: Array<UserAndRolesType>
	): Array<UserAndRolesType> => {
		const usersfilteredByRoles: Array<UserAndRolesType> = []
		let currentUsersRoles: Array<string> = []

		usersAndRoles
			.slice(0)
			.map((user: UserAndRolesType) => {
				currentUsersRoles = []
				user.roles
					.slice(0)
					.map((role) => {
						currentUsersRoles.push(role.name)
					})

					const matchesFilter = roleFiltersChecked.every(element => {
						return currentUsersRoles.includes(element)
					});

					if (matchesFilter) {
						usersfilteredByRoles.push(user)
					}
			});	

		return usersfilteredByRoles
	}

	filterUsersByRoles = (role: string): void => {
		this.setState({ isLoading: true });

		let roleFiltersChecked = this.state.roleFiltersChecked,
		pageNumber = this.state.pageNumber;

		roleFiltersChecked = roleFiltersChecked.includes(role)
			? roleFiltersChecked.filter(
					(roleFilter) => roleFilter !== role,
					)
			: roleFiltersChecked.concat(role);

		const itemsPerPage = Number(this.state.itemsPerPage)
			? Number(this.state.itemsPerPage)
			: this.state.itemsPerPage;

		let usersAndRoles = this.state.websiteUsersAndRoles.usersAndRoles;
		pageNumber = this.state.pageNumber;

		if (roleFiltersChecked.length) {
			usersAndRoles = this.getListOfUsersByFilteredRole(
				roleFiltersChecked,
				usersAndRoles,
			);
		}

		pageNumber = this.pastPageRange(
			itemsPerPage,
			pageNumber,
			this.state.usersAndRoles.length,
		);

		this.setState({ usersAndRoles, isLoading: false, roleFiltersChecked });
	}

	pastPageRange = (
		itemsPerPage: string | number,
		pageNumber: number,
		dataCount: number,
	): number => {
		let pastPageRange = false;

		if (Number(itemsPerPage)) {
			itemsPerPage = Number(itemsPerPage);
			pastPageRange = pageNumber * itemsPerPage >= dataCount;
		}

		if (pastPageRange || itemsPerPage === "all") {
			pageNumber = 1;
		}

		return pageNumber;
	};

	getPaginationText = (
		usersCount: number,
		start: number,
		finish: number,
	): string => {
		const amountPerPage = finish - start;
		const paginationExtract =
		usersCount > amountPerPage ? `${start + 1} to ${finish} of ` : "";

		return `Showing ${paginationExtract}${usersCount} user${
			usersCount === 1 ? "" : "s"
		}`;
	};

	getPaginateStartAndFinishPosition = (
		userCount: number,
		pageNumber: number,
		itemsPerPage: number | string,
	): { start: number; finish: number } => {
		const paginationPositions = {
			start: 0,
			finish: userCount,
		};

		const itemAmountIsNumber = Number(itemsPerPage);

		if (itemAmountIsNumber) {
			paginationPositions.start = (pageNumber - 1) * itemAmountIsNumber;
			paginationPositions.finish =
				paginationPositions.start + itemAmountIsNumber <= userCount
					? paginationPositions.start + itemAmountIsNumber
					: userCount;
		}

		return paginationPositions;
	};

	changeAmount = (e: React.ChangeEvent<HTMLSelectElement>): void => {
		let itemsPerPage = (e.target as HTMLSelectElement).value || 25,
			pageNumber = this.state.pageNumber,
			path = stripMultipleQueries(this.state.path, ["amount", "page"]);

		itemsPerPage = Number(itemsPerPage) ? Number(itemsPerPage) : itemsPerPage;

		pageNumber = this.pastPageRange(
			itemsPerPage,
			pageNumber,
			this.state.usersAndRoles.length,
		);

		path = appendQueryParameter(path, "amount", itemsPerPage.toString());
		path = appendQueryParameter(path, "page", pageNumber);

		this.setState({ itemsPerPage, path, pageNumber }, () => {
			this.props.history.push(path);
		});
	};

	changePage = (e: React.MouseEvent<HTMLAnchorElement>): void => {
		e.preventDefault();

		let pageNumber =
				(e.target as HTMLAnchorElement).getAttribute("data-pager") || 1,
			path = removeQueryParameter(this.state.path, "page");

		if (pageNumber === "previous") {
			pageNumber = this.state.pageNumber - 1;
		}
		if (pageNumber === "next") {
			pageNumber = this.state.pageNumber + 1;
		}

		pageNumber = Number(pageNumber);
		path = appendQueryParameter(path, "page", pageNumber);

		this.setState({ pageNumber, path }, () => {
			this.props.history.push(path);
		});
	};

    render(): JSX.Element {
		const { websiteUsersAndRoles, usersAndRoles, error, isLoading, pageNumber, itemsPerPage, } = this.state;

		const lastBreadcrumb = isLoading ? "Loading service details" : websiteUsersAndRoles.website?.service?.name;

		const paginationPositions = this.getPaginateStartAndFinishPosition(
			usersAndRoles?.length,
			pageNumber,
			itemsPerPage,
		);

		const paginationText = this.getPaginationText(
			usersAndRoles?.length,
			paginationPositions.start,
			paginationPositions.finish,
		);

		const usersPaginated = usersAndRoles?.length
		? usersAndRoles.slice(paginationPositions.start, paginationPositions.finish)
		: usersAndRoles;

        return(
            <>
				<Breadcrumbs>
					<Breadcrumb
							data-qa-sel="breadcrumb-administration-link"
							to="/"
							elementType={Link}
						>
						Administration
					</Breadcrumb>
					<Breadcrumb data-qa-sel="breadcrumb-service-link"
						to="/services" 
						elementType={Link}
						>
						Services
					</Breadcrumb>
					<Breadcrumb>{lastBreadcrumb}</Breadcrumb>
				</Breadcrumbs>
				{!error ? (
					<>
						<PageHeader
							heading={isLoading ? "Service name" : `${websiteUsersAndRoles.website.service.name} (${websiteUsersAndRoles.website.environment.name})`}
						/>

						{isLoading ? (
							<p>Loading...</p>
						) : (
							<Grid>
								<GridItem cols={12} md={3}>
									<FilterBox
										name="Roles"
										filters={this.state.rolesForFilter}
										selected={this.state.roleFiltersChecked}
										onCheckboxChange={this.filterUsersByRoles}
										hideFilterPanelHeading={true}
									/>
								</GridItem>
								<GridItem cols={12} md={9} aria-busy={true} >
									<h2 className={styles.websiteSummary} data-qa-sel="users-returned">{paginationText}</h2>
									<Table style={{ display: "table" }}  data-qa-sel='list-of-users'>
										<thead>
											<tr>
												<th>User name</th>
												<th>Email address</th>
												<th>Roles(s)</th>
												<th>Is staff</th>
											</tr>
										</thead>
										<tbody>
										{usersAndRoles.length ?  (
											usersPaginated
												.slice(0)
												.map((user) => (
													<tr key={user.user.nameIdentifier} className="userRecord">
														<td>
															{user.user.firstName} {user.user.lastName}
														</td>
														<td>
															{user.user.emailAddress}
														</td>
														<td>
															{user.roles
																.slice(0)
																.map((roles, index) => {
																	if (this.state.roleFiltersChecked.includes(roles.name) || this.state.roleFiltersChecked.length === 0) {
																		return <div key={index}>{roles.name}</div>;
																	}
																})}
														</td>
														<td>
															{user.user.isStaffMember ? "Yes" : "No"}
														</td>
													</tr>
											))
										) : (
											<tr>
												<td colSpan={4}>0 results found</td>
											</tr>
										)}
										</tbody>
									</Table>
									<Pagination
										onChangePage={this.changePage}
										onChangeAmount={this.changeAmount}
										itemsPerPage={itemsPerPage}
										consultationCount={usersAndRoles.length}
										currentPage={pageNumber}
									/>
								</GridItem>
							</Grid>
						)}
					</>
				) : (
					<ErrorMessage error={error}></ErrorMessage>
				)}
            </>
        );
    }
}