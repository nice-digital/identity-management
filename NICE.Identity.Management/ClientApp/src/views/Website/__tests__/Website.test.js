import React from "react";
import { render, waitFor, waitForElementToBeRemoved, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { MemoryRouter } from "react-router";
import { Website } from "../Website";
import { Endpoints } from "src/data/endpoints";
import usersAndRoles from "./UsersAndRoles.json";

const consoleErrorReset = console.error;

beforeEach(() => {
	console.error = consoleErrorReset;
});

const server = setupServer(
  rest.get(Endpoints.usersAndRolesByWebsite("1"), (req, res, ctx) => {
    return res(ctx.json(usersAndRoles));
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("should show loading message before data has been loaded", () => {
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {}
		}
	};
	render(<Website  {...websiteProps} />, {wrapper: MemoryRouter});
	const loadingMessage = screen.getByText("Loading...", { selector: "p" });
	expect(loadingMessage).toBeInTheDocument();
});

test("should match the snapshot after data has been loaded", async () => {
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {}
		}
	};
	const {container} = render(<Website {...websiteProps} />, {wrapper: MemoryRouter});	
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	expect(container).toMatchSnapshot();
});

test("should show error message when fetch returns 401 error", async () => {
	console.error = jest.fn();	
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {}
		}
	};
	server.use(
		rest.get(Endpoints.usersAndRolesByWebsite("1"), (req, res, ctx) => {
		return res.once(
				ctx.status(401),
				ctx.json({})
			);
		})
	);
	const {container} = render(<Website {...websiteProps} />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Error"}));
	expect(container).toMatchSnapshot();	
});

test("should show error message when fetch returns 500 error", async () => {
	console.error = jest.fn();	
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {}
		}
	};
	server.use(
		rest.get(Endpoints.usersAndRolesByWebsite("1"), (req, res, ctx) => {
			return res.once(
				ctx.status(500),
				ctx.json({})
			);
		})
	);
	const {container} = render(<Website {...websiteProps} />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Error"}));
	expect(container).toMatchSnapshot();	
});

test("should show no results message when filter returns 0 results", async () => {
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {}
		}
	};
	render(<Website {...websiteProps} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const ownerRoleCheckbox = screen.getByLabelText("Owner");
	userEvent.click(ownerRoleCheckbox);
	const noResultsFoundMessage = await screen.findByText("0 results found", { selector: "td"});
	expect(noResultsFoundMessage).toBeInTheDocument();
});

test("should filter users to all in admins when checkbox is clicked", async () => {
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {}
		}
	};
	render(<Website {...websiteProps} />, {wrapper: MemoryRouter});
	await waitFor(() => screen.getByRole("heading", { name: "Showing 4 users" }));
	let userRoles = screen.queryAllByText(/^(Creator|Editor|Owner|ReadOnly)/i, { selector: "div"});
	expect(userRoles.length).toEqual(6);
	const ownerRoleCheckbox = screen.getByLabelText("Administrator");
	userEvent.click(ownerRoleCheckbox);
	await waitFor(() => screen.getByRole("heading", { name: "Showing 2 users"}));
	userRoles = screen.queryAllByText(/^(Creator|Editor|Owner|ReadOnly)/i, { selector: "div"});
	const adminRoleTableCells = await screen.findAllByText("Administrator", { selector: "div"});
	expect(userRoles.length).toEqual(0);
	expect(adminRoleTableCells.length).toEqual(2);
});

test("should show 25 (default page amount) or less results by default when paginated", async () => {
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {}
		}
	};
	render(<Website {...websiteProps} />, {wrapper: MemoryRouter});	
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	expect(screen.getAllByRole("row").length).toBeLessThanOrEqual(26);
});

test("should go to page 2 when next button is clicked", async () => {
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {},
			search: "?amount=3&page=1"
		},
		history: {
			push: jest.fn(),
		}
	};
	render(<Website {...websiteProps} />, {wrapper: MemoryRouter});	
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	const nextPageButton = screen.getByRole("link", { name: "Go to next page"});
	userEvent.click(nextPageButton);
	const usersListSummary = await screen.findByText("Showing 4 to 4 of 4 users", { selector: "h2" });
	const paginationCounter = screen.getByText("Page 2 of 2");
	expect(usersListSummary).toBeInTheDocument();
	expect(paginationCounter).toBeInTheDocument();	

});

test("should go to first page when page 1 button is clicked", async () => {
	const websiteProps = {
		match: {
			params: { id: 1 },
		},
		location: {
			state: {},
			search: "?amount=1&page=4"
		},
		history: {
			push: jest.fn(),
		}
	};
	render(<Website {...websiteProps} />, {wrapper: MemoryRouter});	
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	let usersListSummary = screen.getByText("Showing 4 to 4 of 4 users", { selector: "h2" });
	let paginationCounter = screen.getByText("Page 4 of 4");
	expect(usersListSummary).toBeInTheDocument();
	expect(paginationCounter).toBeInTheDocument();
	const firstPageButton = screen.getByRole("link", { name: "Go to page 1"});
	userEvent.click(firstPageButton);
	usersListSummary = await screen.findByText("Showing 1 to 1 of 4 users", { selector: "h2" });
	paginationCounter = screen.getByText("Page 1 of 4");
	expect(usersListSummary).toBeInTheDocument();
	expect(paginationCounter).toBeInTheDocument();	
});