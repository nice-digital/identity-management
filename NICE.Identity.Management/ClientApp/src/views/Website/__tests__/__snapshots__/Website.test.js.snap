// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot after data has been loaded 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-service-link"
          href="/services"
        >
          Services
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Identity Admin
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/services",
        "name": "Services"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Identity Admin"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Identity Admin (Live)
    </h1>
  </div>
  <div
    class="grid"
  >
    <div
      data-g="12 md:3"
    >
      <div
        class="filter-panel-hide-heading"
      >
        <form>
          <div
            class="filter-panel"
          >
            <h2
              class="filter-panel__heading"
            >
              <button
                aria-controls="filter-panel-body"
                aria-expanded="true"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="filter-panel__heading-icon filter-panel__heading-icon--expanded"
                  fill="none"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26Z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </h2>
            <div
              aria-hidden="false"
              class="filter-panel__body"
              id="filter-panel-body"
            >
              <div
                class="filter-group"
              >
                <h3
                  class="filter-group__heading"
                >
                  <button
                    aria-controls="group-roles"
                    aria-expanded="true"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="filter-group__heading-icon filter-group__heading-icon--expanded"
                      fill="none"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26Z"
                        fill="currentColor"
                      />
                    </svg>
                    <span
                      id="group-heading-roles"
                    >
                      Roles
                       
                    </span>
                  </button>
                </h3>
                <fieldset
                  aria-hidden="false"
                  class="filter-group__options"
                  id="group-roles"
                >
                  <legend>
                    Roles
                  </legend>
                  <label
                    class="filter-option"
                    for="filter_roles_administrator"
                  >
                    <input
                      id="filter_roles_administrator"
                      name="roles"
                      title="Roles - Administrator"
                      type="checkbox"
                      value="Administrator"
                    />
                    <span
                      class="filter-option__text"
                    >
                      Administrator
                    </span>
                  </label>
                  <label
                    class="filter-option"
                    for="filter_roles_creator"
                  >
                    <input
                      id="filter_roles_creator"
                      name="roles"
                      title="Roles - Creator"
                      type="checkbox"
                      value="Creator"
                    />
                    <span
                      class="filter-option__text"
                    >
                      Creator
                    </span>
                  </label>
                  <label
                    class="filter-option"
                    for="filter_roles_editor"
                  >
                    <input
                      id="filter_roles_editor"
                      name="roles"
                      title="Roles - Editor"
                      type="checkbox"
                      value="Editor"
                    />
                    <span
                      class="filter-option__text"
                    >
                      Editor
                    </span>
                  </label>
                  <label
                    class="filter-option"
                    for="filter_roles_owner"
                  >
                    <input
                      id="filter_roles_owner"
                      name="roles"
                      title="Roles - Owner"
                      type="checkbox"
                      value="Owner"
                    />
                    <span
                      class="filter-option__text"
                    >
                      Owner
                    </span>
                  </label>
                  <label
                    class="filter-option"
                    for="filter_roles_readonly"
                  >
                    <input
                      id="filter_roles_readonly"
                      name="roles"
                      title="Roles - ReadOnly"
                      type="checkbox"
                      value="ReadOnly"
                    />
                    <span
                      class="filter-option__text"
                    >
                      ReadOnly
                    </span>
                  </label>
                </fieldset>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      aria-busy="true"
      data-g="12 md:9"
    >
      <h2
        class="h5 mv--0"
        data-qa-sel="users-returned"
      >
        Showing 4 users
      </h2>
      <table
        class="table"
        data-qa-sel="list-of-users"
        style="display: table;"
      >
        <thead>
          <tr>
            <th>
              User name
            </th>
            <th>
              Email address
            </th>
            <th>
              Roles(s)
            </th>
            <th>
              Is staff
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="userRecord"
          >
            <td>
              Thomas
               
              Holding
            </td>
            <td>
              thomas.holding@example.com
            </td>
            <td>
              <div>
                Editor
              </div>
              <div>
                ReadOnly
              </div>
              <div>
                Administrator
              </div>
            </td>
            <td>
              Yes
            </td>
          </tr>
          <tr
            class="userRecord"
          >
            <td>
              Aisha
               
              Bartlett
            </td>
            <td>
              aisha.bartlett@example.com
            </td>
            <td>
              <div>
                ReadOnly
              </div>
            </td>
            <td>
              Yes
            </td>
          </tr>
          <tr
            class="userRecord"
          >
            <td>
              Theo
               
              Bradford
            </td>
            <td>
              theo.bradford@example.com
            </td>
            <td>
              <div>
                Administrator
              </div>
            </td>
            <td>
              No
            </td>
          </tr>
          <tr
            class="userRecord"
          >
            <td>
              Kristin
               
              Patrick
            </td>
            <td>
              kristin.patrick@example.com
            </td>
            <td>
              <div>
                Editor
              </div>
              <div>
                ReadOnly
              </div>
              <div>
                Creator
              </div>
            </td>
            <td>
              Yes
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="flex flexAlignCenter"
      >
        <div
          class="mr--e"
          data-qa-sel="number-of-results-on-page"
        >
          <label
            class="bold mr--c"
            for="itemsPerPage"
          >
            Items per page
          </label>
          <select
            data-qa-sel="result-on-the-page-index"
            id="itemsPerPage"
            name="itemsPerPage"
          >
            <option
              value="25"
            >
              25
            </option>
            <option
              value="50"
            >
              50
            </option>
            <option
              value="all"
            >
              All
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should show error message when fetch returns 401 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-service-link"
          href="/services"
        >
          Services
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span />
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/services",
        "name": "Services"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {}
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when fetch returns 500 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-service-link"
          href="/services"
        >
          Services
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span />
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/services",
        "name": "Services"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {}
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;
